File: MB-Chat/observability_py/health_snapshot.py
from datetime import UTC, datetime

from .confidence_metrics import calculate_confidence_metrics
from .drift_detector import detect_drift_signals
from .review_metrics import calculate_review_metrics
from .types import HealthSnapshot, ObservabilityEvent


def build_health_snapshot(events: list[ObservabilityEvent]) -> HealthSnapshot:
    drift = detect_drift_signals(events)
    confidence = calculate_confidence_metrics(events)
    review = calculate_review_metrics(events)
    degraded_layers = sorted({signal.layer for signal in drift if signal.severity in {"error", "critical"}})
    return HealthSnapshot(
        status="degraded" if degraded_layers else "unknown",
        providers=sorted({str(event.payload_summary.get("provider_name", "unknown")) for event in events if event.layer == "provider"}),
        confidence_health="degraded" if confidence.uncertainty_avg > 0.7 else "unknown",
        review_pressure="high" if review.pending_reviews > 10 else "medium" if review.pending_reviews > 3 else "low",
        escalation_pressure="high" if review.escalation_count > 10 else "medium" if review.escalation_count > 3 else "low",
        degraded_layers=degraded_layers,  # type: ignore[arg-type]
        drift_alerts=drift,
        created_at=datetime.now(UTC).isoformat(),
    )

File: MetaBrain/observability/health-snapshot.ts
import { calculateConfidenceMetrics } from "./confidence-metrics";
import { detectDriftSignals } from "./drift-detector";
import { calculateReviewMetrics } from "./review-metrics";
import type { HealthSnapshot, ObservabilityEvent } from "./types";

export function buildHealthSnapshot(events: ObservabilityEvent[]): HealthSnapshot {
  const drift = detectDriftSignals(events);
  const confidence = calculateConfidenceMetrics(events);
  const review = calculateReviewMetrics(events);
  const degradedLayers = [...new Set(drift.filter((signal) => signal.severity === "error" || signal.severity === "critical").map((signal) => signal.layer))];
  return {
    status: degradedLayers.length ? "degraded" : "unknown",
    providers: [...new Set(events.filter((event) => event.layer === "provider").map((event) => String(event.payload_summary.provider_name ?? "unknown")))],
    confidence_health: confidence.uncertainty_avg > 0.7 ? "degraded" : "unknown",
    review_pressure: review.pending_reviews > 10 ? "high" : review.pending_reviews > 3 ? "medium" : "low",
    escalation_pressure: review.escalation_count > 10 ? "high" : review.escalation_count > 3 ? "medium" : "low",
    degraded_layers: degradedLayers,
    drift_alerts: drift,
    created_at: new Date().toISOString(),
  };
}
